left sidebar: Create condensed views section

karlstolley opened the pull request:
This PR creates a condensed left-sidebar navigation view (previously, we've used the language _collapsed_, but I think _condensed_ better expresses the intent with these changes).

It extends the project begun in #26807 of presenting the new VIEWS header and its condensed navigation items as a grid. The work done in service of that here should make for relatively quick followup PRs of extending grid-based presentation to the Streams and Direct Messages header, as well as the direct messages rows themselves.

The condensed icons here are duplicated at the template level, which should help make it easier to reason about and modify the condensed and expanded navigation items independent of one another, although a longer-term goal would be to componentize those so as to avoid the code-duplication presented here.

Some of the work in this PR is a continuation or rework of that started in #26035, with some different approaches and in certain cases, quite different code.

Fixes: #25902

[CZO design discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Make.20views.20section.20collapsible.20feedback/near/1661435)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

With placement of redesign icons, as of 2023-10-19

| Light mode | Dark mode |
| --- | --- |
| <img width="500" alt="redesign-icons-light" src="https://github.com/zulip/zulip/assets/170719/3c1b5edb-a896-4d22-8951-3b258e7e2060"> | <img width="500" alt="redesign-icons-dark" src="https://github.com/zulip/zulip/assets/170719/b35672e2-bff6-46a7-b377-29ce01802e66"> |
| Light mode, with dots | Dark mode, with dots |
| <img width="500" alt="redesign-icons-with-unread-dots-light" src="https://github.com/zulip/zulip/assets/170719/a3706e52-4280-4f3c-806d-199fdd4ceaad"> | <img width="500" alt="redesign-icons-with-unread-dots-dark" src="https://github.com/zulip/zulip/assets/170719/ce5b3574-6232-4064-838a-3d85a59310e3"> |



Static appearance, as of 2023-10-18:

| Light mode | Dark mode |
| --- | --- |
| <img width="500" alt="light-mode-2023-10-18" src="https://github.com/zulip/zulip/assets/170719/f078541e-4969-4193-93e3-3f217ce64fe8"> | <img width="500" alt="dark-mode-2023-10-18" src="https://github.com/zulip/zulip/assets/170719/7806edca-dc9c-4a44-9561-04d53584561f"> |

Interactivity, as of 2023-10-18:

| Light mode | Dark mode |
| --- | --- |
| ![light-mode-interactions-2023-10-18](https://github.com/zulip/zulip/assets/170719/9439a7a4-88b8-4f45-b739-e9f4201d789d) | ![dark-mode-interactions-2023-10-18](https://github.com/zulip/zulip/assets/170719/4e64e846-2cbb-46ca-a71c-45e1212864b3) |

The VIEWS heading properly takes an ellipsis in expanded or condensed states, making this safer for internationalization.

| Expanded view, mock long title | Condensed view, mock long title |
| --- | --- |
| <img width="495" alt="wip-long-view-title-expanded" src="https://github.com/zulip/zulip/assets/170719/fbf04de4-57a5-4031-a647-7e49120a4fa9"> | <img width="495" alt="wip-long-view-title-condensed" src="https://github.com/zulip/zulip/assets/170719/c3241708-9203-4618-85cc-7b0ab4b00093"> |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -593,6 +662,88 @@ li.top_left_scheduled_messages {
+    /* Use a next-sibling combinator (+) to use CSS to show and hide
+       filter rows as needed, based on the narrow. */
+    &.showing-condensed-navigation {
+        + #left-sidebar-navigation-list {
```
I suspect that this selector might prove controversial. But 1) even in https://zulip-sidebar.vladkorobov.repl.co/, we expect the header to remain adjacent to the full items, making this less brittle than meets the eye, and 2) I would prefer to avoid the complexity of mucking about in JavaScript with the expanded nav area for things that are happening in the condensed area.

But of course I'm happy to discuss and change.

timabbott replied:
Yeah this seems safe enough; it doesn't seem very likely that someone would be tempted to add something between these elements.

karlstolley commented on web/styles/left_sidebar.css:
```diff
@@ -593,6 +662,88 @@ li.top_left_scheduled_messages {
+        }
+    }
+
+    #toggle-top-left-navigation-area-icon {
```
I'm super happy about the way that the grid I planned in #26807 is applicable here (that said, the condensed views area is basically greenfield material, so _of course_ it works), but I have solved some problems here that were hanging me up on parallel work to apply the grid to existing DM and streams headers.

timabbott commented on web/src/click_handlers.js:
```diff
@@ -715,6 +716,18 @@ export function initialize() {
+        }
+        e.stopPropagation();
+        left_sidebar_navigation_area.toggle_condensed_navigation_area();
+    });
```
We've generally been moving towards not putting anything here that doesn't need to be, in favor of the module itself declaring click handlers in its `initialize` method.

This module is a bit of a historical artifact.

karlstolley replied:
Good to know. Thanks!

timabbott commented:
I gave this a read and have no real concerns with what is present so far; posted a couple small comments.

alya commented:
Playing with the PR, I feel like the three-dot menu feels too hard to discover. We should probably follow one of the two patterns we have in the left sidebar: either always show it, or show it when hovering anywhere over the row. Always showing it seems more consistent with STREAMS and DIRECT MESSAGES, at least in the current design.

alya commented:
All the collapsed icons should have tooltips. The expanded Starred messages and Mentions views currently don't have tooltips because there's no information to add to what's already on the screen, but the name of the view will be helpful in the collapsed state.

karlstolley commented:
> Playing with the PR, I feel like the three-dot menu feels too hard to discover. We should probably follow one of the two patterns we have in the left sidebar: either always show it, or show it when hovering anywhere over the row. Always showing it seems more consistent with STREAMS and DIRECT MESSAGES, at least in the current design.

I agree. I think I hid it because that was the suggestion in the design discussion. But I'm happy to have an always-on state.

karlstolley commented:
> All the collapsed icons should have tooltips. The expanded Starred messages and Mentions views currently don't have tooltips because there's no information to add to what's already on the screen, but the name of the view will be helpful in the collapsed state.

That makes sense. I can add those. The other tooltips in this area are for showing keyboard shortcuts, though. And there currently aren't any for either of those views.

alya commented:
Yeah, that's fine, it's just to give the name of the view if you aren't sure what the icon means.

alya commented:
Something weird is happening with the message count border disappearing on hover, in dark theme only.
<img width="205" alt="Screenshot 2023-10-17 at 12 59 39 PM" src="https://github.com/zulip/zulip/assets/2090066/992b836b-fad0-4ac2-9245-2edbd63f5c27">
<img width="193" alt="Screenshot 2023-10-17 at 12 59 34 PM" src="https://github.com/zulip/zulip/assets/2090066/943e9264-82b7-4311-b078-fb04c81a043b">

alya commented:
I'm finding the unread message dots a bit hard to see in dark theme. Is that intentional? They are much darker than the unread dots elsewhere in the left sidebar.

<img width="172" alt="Screenshot 2023-10-17 at 12 57 49 PM" src="https://github.com/zulip/zulip/assets/2090066/93d5a42c-952c-4f3b-bbcc-7a24b7929894">

<img width="74" alt="Screenshot 2023-10-17 at 12 57 54 PM" src="https://github.com/zulip/zulip/assets/2090066/ba4c176f-ea00-49e3-aaf3-38ab50ce6e9a">

karlstolley commented:
Oooh, nice catches on the dark mode. I've gotten into a bad habit of slacking on checking dark mode lately, which is weird because that's usually how I use CZO...I'll fix.

alya commented:
Some of the icons feel a bit small to me, especially Recent conversations. That one doesn't have an unread dot, so it seems like it could easily be larger?

alya commented:
I think that's all I've got! All pretty tactical tweaks -- I'm excited to get this test-deployed on CZO soon!

karlstolley commented:
Thanks Alya! I'll message back tomorrow when these fixes are up. I really appreciate your review :)

timabbott commented:
I did another skim read; this should be ready to test-deploy once those tweaks are done.

karlstolley commented:
Okay @alya! I believe I've addressed all of your feedback here, plus a few small fixes I noticed myself, including:

* Corrected, consistent noatification-dot colors across light and dark modes
* Corrected hover styles in the Tippy for Drafts and Scheduled messages
* Always-visible vdots, with a new vdots color for use in left-sidebar headings that relates to the existing icon colors
* Larger click-target on the vdots, to match the generous targets on the navigational icons
* Tuned icon sizes for the condensed view, increasing the size of the clock for recents, and decreasing the size of the left-aligned block on all messages (see a question about that below, though)
* Informational tooltips on the @ and star icons (condensed icons only)
* Rounded borders on condensed icons outside the :hover state to make the momentary flash effect when, for example, a new @-mention comes in look better

Updated screenshots for light AND dark mode are at the top of the PR :)

The one question I had is whether we want a background-hover on the vdots. I think it works fine and more consistently with other vdots by highlighting only the dots themselves, and it also makes it clearer that the vdots are informational, not navigational. But I wanted to ask all the same :)

timabbott commented:
One thing I notice is the view hider doesn't seem to show a focus triangle when clicked, while the "DIRECT MESSAGES" one does. 

![image](https://github.com/zulip/zulip/assets/2746074/ef85745d-c00a-41f7-837f-65e97c71e2e9)

I think this entails two bugs with existing "DIRECT MESSAGES" element: (1) we don't want a focus rectangle on click, only on tabbing there, (2) the focus rectangle for "DIRECT MESSAGES" looks bad. 

But there's probably a bug in this new code too, that none of these new elements have a focus rectangle when tabbed to:
![image](https://github.com/zulip/zulip/assets/2746074/262e8d85-4b8c-4cb0-85ab-2b9058831729)

I don't considering this a blocker for merging, in that it's not an area we expect a lot of keyboard usage of, and it's a class of thing we've had before for a long time on elements like the "All direct messages" icon; but might be worth fixing up while we're here and waiting for this to get more QA.

> The one question I had is whether we want a background-hover on the vdots. I think it works fine and more consistently with other vdots by highlighting only the dots themselves, and it also makes it clearer that the vdots are informational, not navigational. But I wanted to ask all the same :)

Agreed, this feels correct to me.

timabbott commented:
Deploying this on chat.zulip.org.

timabbott commented:
Just to flag this, I tested this experience as a logged-out spectator, and it looks like this, which seems totally fine:

![image](https://github.com/zulip/zulip/assets/2746074/2cedf996-3432-4546-aaca-be3f0668793a)

karlstolley commented:
Tim said:

> One thing I notice is the view hider doesn't seem to show a focus triangle when clicked, while the "DIRECT MESSAGES" one does.

Great catch. Yes, let's get both of these working right, care of `:focus-visible`, which should show the focus indicator only in the keyboard case (which we want) and not the pointer case (which we don't want).

timabbott commented:
Merged, thanks @karlstolley and @Ddharmani3!
